src/geom/Endpoint.cc

```diff
@@ -0,0 +1,148 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include <memory>
+#include <vector>
+
+#include "astshim.h"
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/SpherePoint.h"
+#include "lsst/afw/geom/Endpoint.h"
+
+namespace lsst {
+namespace afw {
+namespace geom {
+
+template <typename Point, typename Array>
+std::shared_ptr<ast::Frame> BaseEndpoint<Point, Array>::makeFrame() const {
+    return std::make_shared<ast::Frame>(getNAxes());
+}
+
+template <typename Point>
+std::vector<double> BaseVectorEndpoint<Point>::dataFromPoint(Point const& point) const {
+    const int nAxes = this->getNAxes();
+    std::vector<double> ret(nAxes);
+    for (int axInd = 0; axInd < nAxes; ++axInd) {
+        ret[axInd] = point[axInd];
+    }
+    return ret;
+}
+
+template <typename Point>
+ndarray::Array<double, 2, 2> BaseVectorEndpoint<Point>::dataFromArray(Array const& arr) const {
+    const int nAxes = this->getNAxes();
+    const int nPoints = this->getNPoints(arr);
+    ndarray::Array<double, 2, 2> data = ndarray::allocate(ndarray::makeVector(nPoints, nAxes));
+    auto dataRowIter = data.begin();
+    for (auto const & point: arr) {
+        for (int axInd = 0; axInd < nAxes; ++axInd) {
+            (*dataRowIter)[axInd] = point[axInd];
+        }
+        ++dataRowIter;
+    }
+    return data;
+}
+
+template <typename Point>
+Point BaseVectorEndpoint<Point>::pointFromData(std::vector<double> const& data) const {
+    this->_assertNAxes(this->_getNAxes(data));
+    return Point(data.data());
+}
+
+template <typename Point>
+std::vector<Point> BaseVectorEndpoint<Point>::arrayFromData(ndarray::Array<double, 2, 2> const& data) const {
+    this->_assertNAxes(this->_getNAxes(data));
+    int const nPoints = this->_getNPoints(data);
+    Array arr;
+    arr.reserve(nPoints);
+    for (auto dataRowIter = data.begin(), dataRowEnd = data.end(); dataRowIter != dataRowEnd; ++dataRowIter) {
+        arr.emplace_back(dataRowIter->getData());
+    }
+    return arr;
+}
+
+std::vector<double> GenericEndpoint::dataFromPoint(Point const& point) const {
+    this->_assertNAxes(_getNAxes(point));
+    return point;
+}
+
+ndarray::Array<double, 2, 2> GenericEndpoint::dataFromArray(Array const& arr) const {
+    this->_assertNAxes(_getNAxes(arr));
+    return ndarray::copy(arr);
+}
+
+std::vector<double> GenericEndpoint::pointFromData(std::vector<double> const& data) const {
```

Comment:
std::vector is the generic form used to represent one point of data: one element per axis. I'm not entirely happy with it but the other alternatives I considered are:
- Use a 1-d ndarray. That would require a change to astshim. It may be worth considering because vector is copied and ndarray is not. But it's only a few numbers, and vectors are easier to work with both in C++ and Python -- in Python they can be a list or any collection of numbers.
- Use a 2-d ndarray with only one row. This would work, but it seems clumsy and unnatural since the number of rows must always be 1.